### PR TITLE
[Wallet] Add simplest BIP32/deterministic key generation implementation

### PR DESCRIPTION
**Pull request by jonasschnelli**
Probably most simplest HD implementation.
- New wallets will use hd/bip32 (enabling hd only possible during wallet creation/first start)
- HD can be disabled with `--usehd=0`
- Only hardened key derivation
- Fixed keypath (m/0'/0'/k')
- Only one additional database object (`CHDChain`) which contains the CKeyID for the master key together with the child key counter

I'll add some tests once there are some conceptual acks (and also to keep the diff at a very minimum).

### CODE REVIEW

**Comment by slush0** [comment]
Nice minimalistic PR. One question though. Why to use m/0'/0'/i instead of something more standard? BIP32 proposes m/0'/0/i for external chains (I suppose Core in current implementation does not need internal/change chain). Or even better - use m/0/i, which is forward compatible with BIP44 (maybe Core will make it somedays). m/0/i is subset of BIP44 - an account.

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -3036,6 +3104,7 @@ std::string CWallet::GetWalletHelpString(bool showDebug)
         strUsage += HelpMessageOpt("-sendfreetransactions", strprintf(_("Send transactions as zero-fee transactions if possible (default: %u)"), DEFAULT_SEND_FREE_TRANSACTIONS));
     strUsage += HelpMessageOpt("-spendzeroconfchange", strprintf(_("Spend unconfirmed change when sending transactions (default: %u)"), DEFAULT_SPEND_ZEROCONF_CHANGE));
     strUsage += HelpMessageOpt("-txconfirmtarget=<n>", strprintf(_("If paytxfee is not set, include enough fee so transactions begin confirmation on average within n blocks (default: %u)"), DEFAULT_TX_CONFIRM_TARGET));
+    strUsage += HelpMessageOpt("-usehd", _("Use hierarchical deterministic key generation (HD) after bip32. Has only affect during wallet creation/first start") + " " + strprintf(_("(default: %u)"), DEFAULT_USE_HD_WALLET));
```
Comment (paveljanik) [inline src/wallet/wallet.cpp]: No need for `strprintf`. See previous line.

File: src/wallet/walletdb.cpp [hunk 2]
```diff
@@ -599,6 +599,16 @@ ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
                 return false;
             }
         }
+        else if (strType == "hdchain")
+        {
+            CHDChain chain;
+            ssValue >> chain;
+            if (!pwallet->SetHDChain(chain, true))
+            {
+                strErr = "Error reading wallet database: AddChain failed";
```
Comment (paveljanik) [inline src/wallet/walletdb.cpp]: s/AddChain/SetHDChain/

File: src/wallet/walletdb.cpp [hunk 3]
```diff
@@ -599,6 +599,16 @@ ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
                 return false;
             }
         }
+        else if (strType == "hdchain")
```
Comment (paveljanik) [inline src/wallet/walletdb.cpp]: Repeating `"hdchain"` several times reminds me there was some patch redefining these as constants...

**Comment by paveljanik** [comment]
Concept ACK

Nice, small, elegant 👍

File: src/wallet/wallet.cpp [hunk 4]
```diff
@@ -91,7 +91,45 @@ CPubKey CWallet::GenerateNewKey()
     bool fCompressed = CanSupportFeature(FEATURE_COMPRPUBKEY); // default to compressed public keys if we want 0.6.0 wallets
 
     CKey secret;
-    secret.MakeNewKey(fCompressed);
+
+    // Create new metadata
```
Comment (paveljanik) [inline src/wallet/wallet.cpp]: This line is there two times in the file.

**Comment by paveljanik** [comment]
What should RPC `dumpwallet` dump after this?

**Comment by paveljanik** [comment]
Some import path needed then.

**Comment by laanwj** [comment]
Concept ACK

File: src/wallet/walletdb.cpp (hunk 3 above)
Comment (laanwj) [inline src/wallet/walletdb.cpp]: Which one? We did the redefine-magic-strings-as-constants it in other places, but also for the wallet?

Comment (paveljanik) [inline src/wallet/walletdb.cpp]: Yes, because some of them are repeated several times in the file. But maybe I'm remembering some other patch.

Comment (paveljanik) [inline src/wallet/walletdb.cpp]: Ah, it probably was #5707.

Comment (jonasschnelli) [inline src/wallet/walletdb.cpp]: Yes. I did refactor out these stings into constants serval times. But this should be done in another PR and hopefully after this has been merged.

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: Hmm... I have looked at `walletbroadcast`, there `strprintf` is also used. Isn't more flexible for the translation then?

**Comment by jonasschnelli** [comment]
@slush0: One reason for the current keypath is the use hardened-only key derivation.
I personally think people overestimate the switch-wallet feature. Importing a bip32 master key together with a given keypath can be tricky (lookup window).

But I agree.
If and once this PR gets merged into master, next thing I would do is writing a PR that would make the keypath flexible.

We could discuss the default keypath and the keypath flexibility in a such follow up PR.

This PR only intent to solve the "backup problem" by using deterministic key derivation. So we could even think of using m/k'.

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (MarcoFalke) [inline src/wallet/wallet.cpp]: I think you want to `strprintf(_(string))` the whole string and not two separate stings, as other languages might have different grammar. but consider it a nit.

**Comment by jonasschnelli** [comment]
Fixed @paveljanik nits.

**Comment by slush0** [comment]
> I personally think people overestimate the switch-wallet feature.

@jonasschnelli It's actually quite useful feature, and it may be even for Core. I can imagine somebody who'll have Qt wallet awfully out of sync and will need to get his funds ASAP. Then loading the backup to some other software like Multibit HD will do the job. BIP44 is actually being used by most of current wallets so heading towards the same structure makes sense.

> This PR only intent to solve the "backup problem" 

Fair point and I appreciate somebody is finally addressing it.

File: src/wallet/wallet.cpp [hunk 5]
```diff
@@ -1051,6 +1087,37 @@ CAmount CWallet::GetChange(const CTransaction& tx) const
     return nChange;
 }
 
+bool CWallet::SetHDMasterKey(const CKey& key)
+{
+    LOCK(cs_wallet);
+
+    // store the key as normal "key"/"ckey" object
+    // in the the database
```
Comment (paveljanik) [inline src/wallet/wallet.cpp]: the the

File: src/wallet/wallet.h [hunk 6]
```diff
@@ -57,6 +57,9 @@ static const unsigned int DEFAULT_TX_CONFIRM_TARGET = 2;
 static const unsigned int MAX_FREE_TRANSACTION_CREATE_SIZE = 1000;
 static const bool DEFAULT_WALLETBROADCAST = true;
 
+//! if set, all keys will be deriven by using BIP32
```
Comment (paveljanik) [inline src/wallet/wallet.h]: derived

File: src/wallet/walletdb.cpp (hunk 3 above)
Comment (paveljanik) [inline src/wallet/walletdb.cpp]: Definitely not for this PR, of course.

File: src/wallet/walletdb.h [hunk 7]
```diff
@@ -40,6 +40,34 @@ enum DBErrors
     DB_NEED_REWRITE
 };
 
+/* simple hd chain data model */
+class CHDChain
+{
+public:
+    uint32_t nExternalChainCounter;
+    CKeyID  masterKeyID; /* master key hash160 */
```
Comment (paveljanik) [inline src/wallet/walletdb.h]: two spaces

Comment (MarcoFalke) [inline src/wallet/walletdb.h]: Maybe just clang-format the diff?

**Comment by jonasschnelli** [comment]
Force push fixed @paveljanik nits. Formatted some of the code with clang-format.

File: src/wallet/wallet.cpp [hunk 8]
```diff
@@ -91,7 +91,45 @@ CPubKey CWallet::GenerateNewKey()
...
+        masterKey.SetMaster(key.begin(), key.size());
+
+        // derive m/0'
+        masterKey.Derive(accountKey, 0 | 0x80000000);
+
+        // derive m/0'/0'
+        accountKey.Derive(externalChainChildKey, 0 | 0x80000000);
+
+        // derive child key at next index
+        externalChainChildKey.Derive(childKey, hdChain.nExternalChainCounter | 0x80000000);
```
Comment (sipa) [inline src/wallet/wallet.cpp]: I think you should check whether the derived key already exists in the keystore, and if so, try the next chaincounter.

**Comment by achow101** [comment]
Concept ACK

So as I understand this, it won't affect the keys already in the wallet, right? It just generates all new keys using the HD master key?

Also there should be an RPC command to dump the master private key.

**Comment by jonasschnelli** [comment]
@achow101:
for now (this PR), HD will only be enabled when you create a new wallet (first start).

> Also there should be an RPC command to dump the master private key.

I agree. Have a look at #6265 and #7273.
This is my third try to get HD into Cores wallet. I think the way to go is this extremely simple basic step.
We can add RPC/API changes later.

File: src/wallet/wallet.cpp (hunk 8 above)
Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: I added a loop to derive the next key as long as its not known by the wallet. I think this is required otherwise we might run into an exception if someone restores a backup or has imported a derived key of a higher child index.

File: src/wallet/wallet.h [hunk 9]
```diff
@@ -574,6 +577,9 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
 
     void SyncMetaData(std::pair<TxSpends::iterator, TxSpends::iterator>);
 
+    /* the hd chain data model (internal and external counters) */
```
Comment (instagibbs) [inline src/wallet/wallet.h]: external only for now? (and in the other 2 places in the PR)

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (instagibbs) [inline src/wallet/wallet.cpp]: s/Has only affect/Only has effect/

File: src/wallet/walletdb.h [hunk 10]
```diff
@@ -40,6 +40,35 @@ enum DBErrors
     DB_NEED_REWRITE
 };
 
+/* simple hd chain data model */
+class CHDChain
+{
+public:
+    uint32_t nExternalChainCounter;
+    CKeyID masterKeyID; /* master key hash160 */
```
Comment (MarcoFalke) [inline src/wallet/walletdb.h]: Nit: I think you want to use `//!<` when placing a trailing comment.

File: src/wallet/wallet.cpp [hunk 11]
```diff
@@ -1051,6 +1089,37 @@ CAmount CWallet::GetChange(const CTransaction& tx) const
...
 
+bool CWallet::SetHDMasterKey(const CKey& key)
+{
+    LOCK(cs_wallet);
+
+    // store the key as normal "key"/"ckey" object
+    // in the database
+    // key metadata is not required
+    CPubKey pubkey = key.GetPubKey();
+    if (!AddKeyPubKey(key, pubkey))
```
Comment (instagibbs) [inline src/wallet/wallet.cpp]: nit: simply calling AddKey should work here

**Comment by instagibbs** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/8035/commits/cc1df8fa6f4849b65187f60a3cb4c83da3d16788

Glad to see a really simple PR. Baby steps better than no steps.

**Comment by jonasschnelli** [comment]
Added a squash me commit with some non-code nit fixes.

File: src/wallet/wallet.cpp [hunk 12]
```diff
@@ -91,7 +91,47 @@ CPubKey CWallet::GenerateNewKey()
...
+        CExtKey masterKey;             //hd master key
+        CExtKey accountKey;            //key at m/0'
+        CExtKey externalChainChildKey; //key at m/0'/0'
+        CExtKey childKey;              //key at m/0'/0'/<n>'
+
+        // try to get the master key
+        if (!GetKey(hdChain.masterKeyID, key))
+            throw std::runtime_error("CWallet::GenerateNewKey(): Master key not found");
+
+        assert(key.size() == 32);
```
Comment (pstratem) [inline src/wallet/wallet.cpp]: Why is this an assertion but missing the master key is an exception?

File: src/wallet/wallet.cpp [hunk 13]
```diff
@@ -91,7 +91,47 @@ CPubKey CWallet::GenerateNewKey()
...
+        masterKey.SetMaster(key.begin(), key.size());
+
+        // derive m/0'
+        masterKey.Derive(accountKey, 0 | 0x80000000);
```
Comment (pstratem) [inline src/wallet/wallet.cpp]: Needs comment about hardened derivation.

File: src/wallet/wallet.cpp [hunk 14]
```diff
@@ -91,7 +91,47 @@ CPubKey CWallet::GenerateNewKey()
...
+            // increment childkey index
+            hdChain.nExternalChainCounter++;
+        } while(HaveKey(childKey.key.GetPubKey().GetID()));
+        secret = childKey.key;
+
+        // update the chain model in the database
+        if (!CWalletDB(strWalletFile).WriteHDChain(hdChain))
+            throw std::runtime_error("CWallet::GenerateNewKey(): Writing HD chain model failed");
+    } else
+        secret.MakeNewKey(fCompressed);
```
Comment (pstratem) [inline src/wallet/wallet.cpp]: Please add braces here for the else block.

File: src/wallet/wallet.cpp (hunk 12 above)
Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: I agree. This assertion should be removed.

File: src/wallet/wallet.cpp (hunk 14 above)
Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: our clang format script accepted that.
What's wrong with it? :-)

**Comment by jonasschnelli** [comment]
Added another squash-me commit with minor nit fixes.

**Comment by sipa** [comment]
@jonasschnelli clang-format does not reject anything, it just reformats.

**Comment by jonasschnelli** [comment]
@sipa: right! Sorry. I meant it didn't reformat the `else` part.

**Comment by sipa** [comment]
@jonasschnelli It can't reformat it, because all it does is change whitespace. What we want here is more than just changing whitespace, but there is no way to make clang do that (and also not make it yell about it). Just because a whitespace formatting tool accepts the code, doesn't mean it satisfies all our style guidelines. For example, it also does not check consistency of variable names.

File: src/wallet/wallet.cpp (hunk 14 above)
Comment (MarcoFalke) [inline src/wallet/wallet.cpp]: I think that is what clang-tidy is supposed to do ;)

**Comment by jonasschnelli** [comment]
Amend force pushed last squash-me commit with the `else` + bracket change.

File: src/wallet/wallet.cpp [hunk 15]
```diff
@@ -91,7 +91,48 @@ CPubKey CWallet::GenerateNewKey()
...
+
+        // try to get the master key
+        if (!GetKey(hdChain.masterKeyID, key))
+            throw std::runtime_error("CWallet::GenerateNewKey(): Master key not found");
+
+        masterKey.SetMaster(key.begin(), key.size());
+
+        // derive m/0'
+        // use hardened derivation (child keys > 0x80000000 are hardened after bip32)
+        masterKey.Derive(accountKey, 0 | 0x80000000);
```
Comment (jtimon) [inline src/wallet/wallet.cpp]: Sorry, I don't understand this ```| 0x80000000``` from the comments...

Comment (instagibbs) [inline src/wallet/wallet.cpp]: Hardened keys are >= 0x80000000, not > according to bip32

Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: It's specified here:
https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#child-key-derivation-ckd-functions

You can derive 2^32 key, child keys <= 2147483648 (<=0x80000000) are non-hardened key. Derived child keys > 2147483648 are hardened keys.

This line:
`masterKey.Derive(accountKey, 0 | 0x80000000);`
.. is equal to ..
`masterKey.Derive(accountKey, 0 + 2147483648)`

Comment (MarcoFalke) [inline src/wallet/wallet.cpp]: I think the comment has a typo?

**Comment by jtimon** [comment]
I still have limited knowledge of the wallet, but thanks to this being so simple I was able to review it and everything looks good.
something-between-concept-and-ut ACK

File: src/wallet/wallet.cpp (hunk 15 above)
Comment (jtimon) [inline src/wallet/wallet.cpp]: Oh, I forgot about this, thanks. IIRC, pycoin (what I've used for bip32 so far) used an asterisk in its interface for hardened keys. I wouldn't complain if the link was added to the comment, but maybe an appropriately named constant for  0x80000000 would be enough.

**Comment by arowser** [comment]
Can one of the admins verify this patch?

**Comment by jonasschnelli** [comment]
Rebased. Squashed squash-me commits.

**Comment by fanquake** [comment]
This needs to be added to doc/bips.md. I can add it to #8121 if you'd rather not add additional commits.

**Comment by paveljanik** [comment]
Doesn't build here:
```
wallet/wallet.cpp:3219:9: error: use of undeclared identifier 'RandAddSeedPerfmon'
```

See #7891, it should be enough to remove it.

**Comment by MarcoFalke** [comment]
I don't think this should be added to doc/bips.md, as this pull does not add any interface at all. Though, the release notes should be updated.

File: src/wallet/wallet.cpp [hunk 16]
```diff
@@ -91,7 +91,48 @@ CPubKey CWallet::GenerateNewKey()
...
+        CExtKey childKey;              //key at m/0'/0'/<n>'
+
+        // try to get the master key
+        if (!GetKey(hdChain.masterKeyID, key))
+            throw std::runtime_error("CWallet::GenerateNewKey(): Master key not found");
+
+        masterKey.SetMaster(key.begin(), key.size());
+
+        // derive m/0'
+        // use hardened derivation (child keys > 0x80000000 are hardened after bip32)
```
Comment (MarcoFalke) [inline src/wallet/wallet.cpp]: comment-nit: It is >= 0x80000000

**Comment by jonasschnelli** [comment]
Fixed and force pushed the rebase issue `RandAddSeedPerfmon` (#7891).

File: src/wallet/wallet.cpp (hunk 15 above)
Comment (sipa) [inline src/wallet/wallet.cpp]: Agree about adding a named constant.

File: src/wallet/wallet.cpp (hunk 11 above)
Comment (sipa) [inline src/wallet/wallet.cpp]: I'd like to know why this comment does not apply.

**Comment by sipa** [comment]
utACK, will test.

**Comment by jonasschnelli** [comment]
Added a commit that...
* ... uses a constant for 0x80000000 (hardened child index range limit)
* ... extend the source code comment
* ... simplify `0 | 0x80000000` in `0x80000000` (and is now a const)

**Comment by laanwj** [comment]
This is a high-profile feature. Definitely needs mention in release notes.

**Comment by jonasschnelli** [comment]
Okay. I'll write something for the release notes.

**Comment by jonasschnelli** [comment]
Added a part in the `release-notes.md` (thanks for lang review).
Updated the `bips.md`.

File: doc/release-notes.md [hunk 17]
```diff
@@ -112,6 +112,24 @@ feerate. [BIP 133](https://github.com/bitcoin/bips/blob/master/bip-0133.mediawik
 
 ### Wallet
 
+Hierarchical Deterministic Key Generation
+-----------------------------------------
+Newly created wallets will use hierarchical deterministic key generation
+according to BIP32 (keypath m/0'/0'/k').
+Existing wallets will still use traditional key generation.
+
+Backups, regardless of when they have been created, can therefore be
```
Comment (MarcoFalke) [inline doc/release-notes.md]: Nit: Let's call it "Backups of HD wallets"?

File: doc/release-notes.md [hunk 18]
```diff
@@ -112,6 +112,24 @@ feerate. [BIP 133](https://github.com/bitcoin/bips/blob/master/bip-0133.mediawik
...
+used to re-generate all possible private keys, even the ones who hasn't
```
Comment (MarcoFalke) [inline doc/release-notes.md]: "even the ones **which haven't** already"?

File: src/wallet/wallet.cpp [hunk 19]
```diff
@@ -3058,6 +3132,7 @@ std::string CWallet::GetWalletHelpString(bool showDebug)
         strUsage += HelpMessageOpt("-sendfreetransactions", strprintf(_("Send transactions as zero-fee transactions if possible (default: %u)"), DEFAULT_SEND_FREE_TRANSACTIONS));
     strUsage += HelpMessageOpt("-spendzeroconfchange", strprintf(_("Spend unconfirmed change when sending transactions (default: %u)"), DEFAULT_SPEND_ZEROCONF_CHANGE));
     strUsage += HelpMessageOpt("-txconfirmtarget=<n>", strprintf(_("If paytxfee is not set, include enough fee so transactions begin confirmation on average within n blocks (default: %u)"), DEFAULT_TX_CONFIRM_TARGET));
+    strUsage += HelpMessageOpt("-usehd", _("Use hierarchical deterministic key generation (HD) after bip32. Only has effect during wallet creation/first start") + " " + strprintf(_("(default: %u)"), DEFAULT_USE_HD_WALLET));
```
Comment (MarcoFalke) [inline src/wallet/wallet.cpp]: Nit: Might as well call this `-createhdwallet` or similar. Using a lengthy and more verbose option does not hurt if it is expected to be changed from its default only very rarely.

Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: Agree that `-createhdwallet` would make more sense.

**Comment by sipa** [comment]
I prefer `-usehd`, but adding a check that if it is given in case there already exists a wallet, we verify that it matches that wallet. That way you won't ever get any surprises, just set `-usehd`, and you're sure you'll always use HD derivation, set `-usehd=0` and you're sure you'll never.

**Comment by MarcoFalke** [comment]
@sipa If you add the check that your wallet must match with `-usehd` you can't set the default to true, I assume. Otherwise the a user switching from 0.12.1 to 0.13 would see a InitError or InitWarning pop up on every start.

**Comment by sipa** [comment]
If -usehd is not explicitly given, default to whatever is the current
wallet (or true for new wallets). if it is explicitly given, it must match
the existing wallet.

**Comment by MarcoFalke** [comment]
Ok makes sense. (But that is different from txindex, as I remember you can't set -txindex=1 once and then remove it, to make it default to true. I may be mistaken but I should better not hijack this pull's discussion.)

**Comment by instagibbs** [comment]
@MarcoFalke it's slightly different, yes, but txindex behavior wouldn't make sense since you can't "reindex" your HD wallet to legacy.

**Comment by jonasschnelli** [comment]
Added a commit.
Now the PR detects if a user likes to disabled hd (with `-usehd=0`) when using a hd wallet.dat file.
The PR also detects if a user likes to enable hd (with `-usehd=1`) when using a non hd wallet.dat file.

The node will shutdown/stop init in a such mismatch situation.

**Comment by MarcoFalke** [comment]
utACK 71ff55fcc2245e03c786d7e0312755e812d9de8e

File: src/wallet/wallet.cpp [hunk 20]
```diff
@@ -3236,6 +3236,14 @@ bool CWallet::InitLoadWallet()
 
         walletInstance->SetBestChain(chainActive.GetLocator());
     }
+    else
```
Comment (laanwj) [inline src/wallet/wallet.cpp]: Can be simplified a bit:
```
else if (mapArgs.count("-usehd")) {
    bool useHd = GetBoolArg("-usehd", DEFAULT_USEHD);
    if (!walletInstance->hdChain.masterKeyID.IsNull() && !useHD)
        return InitError(strprintf(_("Error loading %s: You can't disable HD on a already existing HD wallet"), walletFile));
    if (walletInstance->hdChain.masterKeyID.IsNull() && useHD)
        return InitError(strprintf(_("Error loading %s: You can't enable HD on a already existing non-HD wallet"), walletFile));
}
```

Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: Thanks. Much simpler.
Force push updated last commit.

**Comment by laanwj** [comment]
utACK afcd77e

Even though this wallet change is reasonably risky, I think this has received enough review for correctness, and it should be merged now to make sure it gets more actual testing.

**Comment by gmaxwell** [comment]
Post merge utACK. Will also test.

**Comment by schildbach** [comment]
Just found out about this BIP. Would like to clarify the keypath: It's purpose number is `0'` not `0`? (`(m/0'/0'/k')` not `(m/0/0'/k')`)

I'm asking because otherwise it could be incompatible to how bitcoinj derives addresses.

**Comment by jonasschnelli** [comment]
@schildbach: We have decided to go with a "hardened only derivation" version for the first implementation. Using `m/0/0'/k'` would introduce the risk of revealing extended pubkey `m/0` together with a non extended child privatekey resulting in revealing the whole chain of keys.

On top of that, I personally think importing a xpriv into another wallet in order to "reconstruct everything" is not a very good concept.

**Comment by schildbach** [comment]
@jonasschnelli Yes that's fine. I just wanted to make sure there is no "partial overlap" with the bitcoinj impl.

**Comment by rebroad** [comment]
So, how does one import a BIP32 master key into an existing wallet?

**Comment by diegoviola** [comment]
I'm switching to Core (from Electrum) after this lands in 0.13, thanks for your efforts.

**Comment by laanwj** [comment]
> So, how does one import a BIP32 master key into an existing wallet?

That functionality currently doesn't exist. Please read the OP: this is the smallest possible change to introduce BIP32 into Bitcoin Core. This made it realistic to get it reviewed and merged in a small timespan. More functionality can be added later.

File: src/wallet/wallet.cpp [hunk 21]
```diff
@@ -1049,6 +1092,37 @@ CAmount CWallet::GetChange(const CTransaction& tx) const
...
+bool CWallet::SetHDMasterKey(const CKey& key)
+{
+    LOCK(cs_wallet);
+
+    // store the key as normal "key"/"ckey" object
+    // in the database
+    // key metadata is not required
+    CPubKey pubkey = key.GetPubKey();
+    if (!AddKeyPubKey(key, pubkey))
+        throw std::runtime_error("CWallet::GenerateNewKey(): AddKey failed");
```
Comment (MarcoFalke) [inline src/wallet/wallet.cpp]: Nit: Wrong function name

**Comment by TheBlueMatt** [comment]
Post-merge review comments, sorry :/ (continued from IRC):
 * The use of an ECDSA private key (stored as such) as the master seed is somewhat strange...While not strictly an issue, I'm concerned about the ability of users to dump this as a regular key, potentially causing some interesting behavior.
* The way I read the current code, every time we go to GenerateNewKey() we will iterate over every key we have already generated before generating a new one (as we never call SetHDChain to store the updated counter on disk)....anyone with a test hd wallet lying around want to see what their disk-counter is?
 * I dont see any discussion of the non-switch of FEATURE_LATEST on the pr....this might complicate future wallet changes...it seems to me that this locks us in such that the next change to wallet version must either be the same non-default or we will, at that time, have to force hd wallets to be the default. Again, I'm not strictly aginst this, just noting that it seems it was not at all discussed but locks us into a future path.
 * I think IsKeyType should be changed to include hdchain, not add an explicit hdchain check in ::Recover, since we probably definitely want it in ::LoadWallet as well.
 * The duplication of the old CTransaction READWRITE(this->nVersion); nVersion = this->nVersion; logic seems strange to me. I always found that a kinda funny way to do things and was glad to see it go in SegWit...Its "hidden functionality" and I'd prefer if we handled that very explicitly instead of via this strange hack.

I think all but the first could easily be fixed even now, if we decide they are, indeed, issues.

**Comment by jonasschnelli** [comment]
Thanks for the post-merge review Matt.

> The use of an ECDSA private key (stored as such) as the master seed is somewhat strange...While not strictly an issue, I'm concerned about the ability of users to dump this as a regular key, potentially causing some interesting behavior.

Yes. Agree. This was done after two of my former PR did not made it into the master (https://github.com/bitcoin/bitcoin/pull/6265 https://github.com/bitcoin/bitcoin/pull/7273) and I was looking for a simpler solutions.
There is a PR to identify the seed when dumping (https://github.com/bitcoin/bitcoin/pull/8206).

> The way I read the current code, every time we go to GenerateNewKey() we will iterate over every key we have already generated before generating a new one (as we never call SetHDChain to store the updated counter on disk)....anyone with a test hd wallet lying around want to see what their disk-counter is?

This is to ensure we don't re-generate an already existing key (there are some cases where this would be possible otherwise).
The iteration should be painless. It's a `std::map` `count()` of an in memory map of all the key.

**Comment by TheBlueMatt** [comment]
re: GenerateNewKey(): Not sure what your response was about (seems to be communication failure there), but it was pointed out to me that I missed the save line and was incorrect.

**Comment by zihad944** [comment]
I foget my sceond wallet passwor. Thats why i cant get backup. I need help

**Comment by jeandudey** [comment]
@zihad944 what problem do you have specifically, what do you mean by "my second wallet password"?

**Comment by zihad944** [comment]
I use sceons password in my block chain. but i forget it.how i  cant send money. I cant see my backu.

**Review by sorawee1970 — COMMENTED** [review]
src/wallet/walletdb.h
